cmd/util/ledger/migrations/filter_unreferenced_slabs_migration_test.go

```diff
@@ -0,0 +1,232 @@
+package migrations
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/onflow/atree"
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/reporters"
+	"github.com/onflow/flow-go/cmd/util/ledger/util"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/convert"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+func TestFilterUnreferencedSlabs(t *testing.T) {
+	t.Parallel()
+
+	// Arrange
+
+	const chainID = flow.Emulator
+	chain := chainID.Chain()
+
+	testFlowAddress, err := chain.AddressAtIndex(1_000_000)
+	require.NoError(t, err)
+
+	testAddress := common.Address(testFlowAddress)
+
+	payloads := map[flow.RegisterID]*ledger.Payload{}
+
+	payloadsLedger := util.NewPayloadsLedger(payloads)
+
+	storageIndices := map[string]uint64{}
+	payloadsLedger.AllocateStorageIndexFunc = func(owner []byte) (atree.SlabIndex, error) {
+		var index atree.SlabIndex
+
+		storageIndices[string(owner)]++
+
+		binary.BigEndian.PutUint64(
+			index[:],
+			storageIndices[string(owner)],
+		)
+
+		return index, nil
+	}
+
+	storage := runtime.NewStorage(payloadsLedger, nil)
+
+	// {Int: Int}
+	dict1StaticType := interpreter.NewDictionaryStaticType(
+		nil,
+		interpreter.PrimitiveStaticTypeInt,
+		interpreter.PrimitiveStaticTypeInt,
+	)
+
+	inter, err := interpreter.NewInterpreter(
+		nil,
+		nil,
+		&interpreter.Config{
+			Storage: storage,
+		},
+	)
+	require.NoError(t, err)
+
+	dict1 := interpreter.NewDictionaryValueWithAddress(
+		inter,
+		interpreter.EmptyLocationRange,
+		dict1StaticType,
+		testAddress,
+	)
+
+	// Storage another dictionary, with a nested array, in the account.
+	// It is not referenced through a storage map though.
+
+	arrayStaticType := interpreter.NewVariableSizedStaticType(nil, interpreter.PrimitiveStaticTypeInt)
+
+	dict2StaticType := interpreter.NewDictionaryStaticType(
+		nil,
+		interpreter.PrimitiveStaticTypeInt,
+		arrayStaticType,
+	)
+
+	dict2 := interpreter.NewDictionaryValueWithAddress(
+		inter,
+		interpreter.EmptyLocationRange,
+		dict2StaticType,
+		testAddress,
+	)
+
+	// NOTE: InsertWithoutTransfer isn't available in Cadence v0.42.
+	//       Number of payloads is increased by 1 by using Insert().
+	dict2.Insert(
+		inter, interpreter.EmptyLocationRange,
+		interpreter.NewUnmeteredIntValueFromInt64(2),
+		interpreter.NewArrayValue(
+			inter,
+			interpreter.EmptyLocationRange,
+			arrayStaticType,
+			testAddress,
+			interpreter.NewUnmeteredIntValueFromInt64(3),
+		),
+	)
```

Comment:
`Insert` transfers the inserted value to the target address. Given it's already created in the account, that leads to the array being stored twice.

We'll want to create the array on the stack, then insert:

```suggestion
	array := interpreter.NewArrayValue(
		inter,
		interpreter.EmptyLocationRange,
		arrayStaticType,
		common.ZeroAddress,
		interpreter.NewUnmeteredIntValueFromInt64(3),
	)

	dict2.Insert(
		inter, interpreter.EmptyLocationRange,
		interpreter.NewUnmeteredIntValueFromInt64(2),
		array,
	)
```

As a result, `totalSlabCount` is correctly 5 again.

However, even after adjusting `totalSlabCount`, the test still fails, slab $3 isn't removed anymore.

Looking at the slabs before and after the migration, this is maybe because the array is inlined into the dictionary now? The idea with this test case preparation here was that we would have two slabs that are unreferenced and need to be filtered out.

Reply:
Good idea to add some elements to the array to ensure it's large enough to be stored in a separate slab and not get inlined 👍  Added a comment explaining why this is done/needed with [4c26d2f](https://github.com/onflow/flow-go/pull/5758/commits/4c26d2f1d08b516d13d9fbba4f8c28e7cdc729ce)